cirq/sim/wave_function.py

```diff
@@ -26,6 +28,100 @@
+    the standard Kronecker convention of numpy.kron.
+
+    Args:
+        state: A sequence representing a wave function in which
+            the ordering mapping to qubits follows the standard Kronecker
+            convention of numpy.kron.
+        index: index of qubit who's bloch vector we want to find.
+            follows the standard Kronecker convention of numpy.kron.
+
+    Returns:
+        A length 3 numpy array representing the qubit's bloch vector.
+
+    Raises:
+        ValueError: if the size of state is not a power of 2.
+        ValueError: if the size of the state represents more than 25 qubits.
+        IndexError: if index is out of range for the number of qubits
+            corresponding to the state.
+    """
+    rho = density_matrix_from_state_vector(state, [index])
+    v = np.zeros(3, dtype=np.float32)
+    v[0] = 2*np.real(rho[0][1])
+    v[1] = 2*np.imag(rho[1][0])
+    v[2] = np.real(rho[0][0] - rho[1][1])
+
+    return v
+
+
+def density_matrix_from_state_vector(state: Sequence,
+    indices: List[int] = None) -> np.ndarray:
+    """Returns the density matrix of the wavefunction.
+
+    Calculate the density matrix for the system on the given qubit
+    indices, with the qubits not in indices that are present in state
+    traced out. If indices is None the full density matrix for state
+    is returned. We assume state follows the standard Kronecker
+    convention of numpy.kron.
+
+    For example:
+        state = np.array([1/np.sqrt(2), 1/np.sqrt(2)], dtype=np.complex64)
+        indices = None
+        gives us \rho = \begin{bmatrix}
+                            0.5 & 0.5
+                            0.5 & 0.5
+                        \end{bmatrix}
+
+    Args:
+        state: A sequence representing a wave function in which
+            the ordering mapping to qubits follows the standard Kronecker
+            convention of numpy.kron.
+        indices: list containing indices for qubits that you would like
+            to include in the density matrix (i.e.) qubits that WON'T
+            be traced out. follows the standard Kronecker convention of
+            numpy.kron.
+
+    Returns:
+        A numpy array representing the density matrix.
+
+    Raises:
+        ValueError: if the size of state is not a power of 2.
+        ValueError: if the size of the state represents more than 25 qubits.
+        IndexError: if the indices are out of range for the number of qubits
+            corresponding to the state.
+    """
+    n_qubits = _validate_num_qubits(state)
+
+    if n_qubits > 26:
+        # coverage: ignore
+        raise ValueError("cannot subscript more than 26 qubit wavefunctions\
+            with np.einsum.")
+
+    if indices is None:
+        return np.outer(state, np.conj(state))
+
+    _validate_indices(n_qubits, indices)
+
+    state = np.asarray(state).reshape((2,)*n_qubits)
+
+    upper_str = np.array([i for i in string.ascii_uppercase[:n_qubits]])
```

Comment:
Let's fix the limitation to 26 by passing in tuples of ints instead of a string, and using only the integers we need.

Use the indices in 0..n to indicate the input axes, but instead of using n..2n for the kept axes, only use n..n+k where k is the number of kept axes. Something like:

```python
k = len(indices)
n = n_qubits
inp = list(range(n))
inp2 = list(inp)
next_index = n
for i in indices:
    inp2[i] = next_index
    next_index += 1
kept = list(range(n, n+k))
out = indices + kept
```

I probably got a bunch of ordering related things backwards in that code. Hopefully it gets across the general idea.

This will replace the qubits <= 26 constraint with qubits + kept <= 52, which is more reasonable. I think we can reasonably assume that kept is less than 17 and qubits is less than 34, since those both require 128 GiB of memory to store and it would take hours to actually extract the value. And even if computers got a lot faster in the far future, numpy would remove the constraint and we'd inherit that.

I wouldn't even bother checking if the user is over the limit. Just let numpy fail when they ask for something ridiculous, so they can complain to the right person.

Reply:
Thank you for the feedback, I've updated the logic now using integer indexing in einsum, what do you think ?